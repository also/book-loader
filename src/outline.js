const cheerio = require('cheerio');


module.exports = function transformToc(toc) {
  const $ = cheerio.load(toc);

  function readToc(elt) {
    if (elt.type === 'comment') {
      return [];
    }

    const children = [];

    const $elt = $(elt);
    $elt.contents().each((i, child) => {
      children.push(...readToc(child));
    });

    const simpleChildren = children.every(({simple}) => simple);
    const listChildren = children.some(({list}) => list);
    let html = simpleChildren ? $elt.html() || elt.data : null;

    function result(opts) {
      const base = {type: opts.type || elt.tagName, list: listChildren};
      if (!simpleChildren) {
        base.children = children;
      } else {
        base.html = html;
      }

      return [Object.assign(base, opts)];
    }

    if (elt.tagName === 'p') {
      return children;
    } else if (elt.type === 'text') {
      if (html === '\n') {
        return [];
      }
      return result({
        type: 'text',
        simple: true
      });
    } else if (elt.tagName === 'ul') {
      return result({
        type: 'list',
        list: true,
      });
    } else if (elt.tagName === 'li') {
      if (children.length === 0) {
        // wat
        return [];
      } else {
        // the title is all the non-list children at the beginning
        let firstListIndex = children.findIndex(({list}) => list);
        if (firstListIndex === -1) {
          firstListIndex = children.length;
        }
        const title = children.slice(0, firstListIndex);
        const listChildren = children.slice(firstListIndex);
        const entries = [];
        listChildren.forEach((child) => {
          if (child.type === 'list') {
            entries.push(...child.children);
          } else {
            throw new Error(`Don't know what to do with children ${$elt.html()}`);
          }
        });
        return result({
          type: 'entry',
          title,
          children: entries
        });
      }
    } else if (elt.tagName === 'a') {
      return result({
        attrs: elt.attribs
      });
    } else {
      return result({
        simple: simpleChildren,
        attrs: elt.attribs
      });
    }
  }
  const outline = readToc($.root().get(0))[0];
  return outline;
};
