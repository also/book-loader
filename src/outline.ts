type Node = {
  type: string;
  children: Node[];
  simple?: boolean;
  list: boolean;
  elt: any;
};

export function transformToc($) {
  function readToc(elt): Node[] {
    if (elt.type === 'comment') {
      return [];
    }

    const children: Node[] = [];

    const $elt = $(elt);
    $elt.contents().each((i, child) => {
      children.push(...readToc(child));
    });

    const simpleChildren = children.every(({ simple }) => !!simple);
    const listChildren = children.some(({ list }) => list);
    let html = simpleChildren ? $elt.html() || elt.data : null;

    function result(opts): Node[] {
      const base: any = { type: opts.type || elt.tagName, list: listChildren };
      // elt is non-enumerable so the result is easy to stringify
      Object.defineProperty(base, 'elt', { enumerable: false, value: elt });
      if (!simpleChildren) {
        base.children = children;
      } else {
        base.html = html;
      }

      return [Object.assign(base, opts)];
    }

    if (elt.tagName === 'p') {
      return children;
    } else if (elt.type === 'text') {
      if (html === '\n') {
        return [];
      }
      return result({
        type: 'text',
        simple: true,
      });
    } else if (elt.tagName === 'ul') {
      return result({
        type: 'list',
        list: true,
      });
    } else if (elt.tagName === 'li') {
      if (children.length === 0) {
        // wat
        return [];
      } else {
        // the title is all the non-list children at the beginning
        let firstListIndex = children.findIndex(({ list }) => list);
        if (firstListIndex === -1) {
          firstListIndex = children.length;
        }

        const title = children.slice(0, firstListIndex);
        const titleElts = title.map(({ elt }) => elt);

        const titleText = $.text(titleElts)
          .replace(/\s+/g, ' ')
          .trim();
        // TODO this is a hack to get cheerio to find an a tag at the root or
        // a child of any of the titleElts. It normally only searches children.
        const url = $('a', [{ children: titleElts }]).attr('href');

        const listChildren = children.slice(firstListIndex);
        const entries: Node[] = [];
        listChildren.forEach(child => {
          if (child.type === 'list') {
            entries.push(...child.children);
          } else {
            throw new Error(
              `Found unexpected "${
                child.elt.tagName
              }" tag after list in "${titleText}": ${$.html(child.elt)}`
            );
          }
        });

        return result({
          type: 'entry',
          title,
          titleText,
          url,
          children: entries,
        });
      }
    } else if (elt.tagName === 'a') {
      return result({
        attrs: elt.attribs,
      });
    } else {
      return result({
        simple: simpleChildren,
        attrs: elt.attribs,
      });
    }
  }
  const outline = readToc($.root().get(0))[0];
  return outline;
}

export function generateBreadcrumbs(outline) {
  const result = new Map();
  const recurse = (node, path) => {
    if (node.type === 'entry') {
      path = [...path, { title: node.titleText, url: node.url }];
      if (node.url != null) {
        const entries = result.get(node.url) || [];
        entries.push(path);
        result.set(node.url, entries);
      }
    }
    if (node.children) {
      node.children.forEach(child => recurse(child, path));
    }
  };

  recurse(outline, []);

  return result;
}
