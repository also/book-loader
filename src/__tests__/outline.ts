const cheerio = require('cheerio');
const markdownIt = require('markdown-it');
import { transformToc } from '../outline';

function createHtml(str, options = {}) {
  return cheerio.load(markdownIt(options).render(str));
}

it('parses simple outline', () => {
  expect(
    transformToc(
      createHtml(
        `
* [One](one.md)
* [Two](two.md)
`
      )
    )
  ).toMatchSnapshot();
});

it('parses tag in link', () => {
  expect(
    transformToc(
      createHtml(
        `
* [*One*](one.md)
* [Two](two.md)
`
      )
    )
  ).toMatchSnapshot();
});

it('parses link in tag', () => {
  expect(
    transformToc(
      createHtml(
        `
* *[One](one.md)*
* [Two](two.md)
`
      )
    )
  ).toMatchSnapshot();
});

it('parses nested list', () => {
  expect(
    transformToc(
      createHtml(
        `
* [One](one.md)
  * [One A](one-a.md)
  * [One b](one-b.md)
* [Two](two.md)
`
      )
    )
  ).toMatchSnapshot();
});

it('ignores comments', () => {
  expect(
    transformToc(
      createHtml(
        `
* [One](one.md)
* [Two](two.md)

<!-- ignore me -->
`,
        { html: true }
      )
    )
  ).toMatchSnapshot();
});
