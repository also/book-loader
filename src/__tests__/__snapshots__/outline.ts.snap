// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignores comments 1`] = `
Object {
  "attrs": Object {},
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "one.md",
              },
              "html": "One",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "One",
          "type": "entry",
          "url": "one.md",
        },
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "two.md",
              },
              "html": "Two",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "Two",
          "type": "entry",
          "url": "two.md",
        },
      ],
      "list": true,
      "type": "list",
    },
  ],
  "list": true,
  "simple": false,
  "type": "root",
}
`;

exports[`parses link in tag 1`] = `
Object {
  "attrs": Object {},
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {},
              "children": Array [
                Object {
                  "attrs": Object {
                    "href": "one.md",
                  },
                  "html": "One",
                  "list": false,
                  "type": "a",
                },
              ],
              "list": false,
              "simple": false,
              "type": "em",
            },
          ],
          "titleText": "One",
          "type": "entry",
          "url": "one.md",
        },
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "two.md",
              },
              "html": "Two",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "Two",
          "type": "entry",
          "url": "two.md",
        },
      ],
      "list": true,
      "type": "list",
    },
  ],
  "list": true,
  "simple": false,
  "type": "root",
}
`;

exports[`parses nested list 1`] = `
Object {
  "attrs": Object {},
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "list": false,
              "title": Array [
                Object {
                  "attrs": Object {
                    "href": "one-a.md",
                  },
                  "html": "One A",
                  "list": false,
                  "type": "a",
                },
              ],
              "titleText": "One A",
              "type": "entry",
              "url": "one-a.md",
            },
            Object {
              "children": Array [],
              "list": false,
              "title": Array [
                Object {
                  "attrs": Object {
                    "href": "one-b.md",
                  },
                  "html": "One b",
                  "list": false,
                  "type": "a",
                },
              ],
              "titleText": "One b",
              "type": "entry",
              "url": "one-b.md",
            },
          ],
          "list": true,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "one.md",
              },
              "html": "One",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "One",
          "type": "entry",
          "url": "one.md",
        },
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "two.md",
              },
              "html": "Two",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "Two",
          "type": "entry",
          "url": "two.md",
        },
      ],
      "list": true,
      "type": "list",
    },
  ],
  "list": true,
  "simple": false,
  "type": "root",
}
`;

exports[`parses simple outline 1`] = `
Object {
  "attrs": Object {},
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "one.md",
              },
              "html": "One",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "One",
          "type": "entry",
          "url": "one.md",
        },
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "two.md",
              },
              "html": "Two",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "Two",
          "type": "entry",
          "url": "two.md",
        },
      ],
      "list": true,
      "type": "list",
    },
  ],
  "list": true,
  "simple": false,
  "type": "root",
}
`;

exports[`parses tag in link 1`] = `
Object {
  "attrs": Object {},
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "one.md",
              },
              "html": "<em>One</em>",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "One",
          "type": "entry",
          "url": "one.md",
        },
        Object {
          "children": Array [],
          "list": false,
          "title": Array [
            Object {
              "attrs": Object {
                "href": "two.md",
              },
              "html": "Two",
              "list": false,
              "type": "a",
            },
          ],
          "titleText": "Two",
          "type": "entry",
          "url": "two.md",
        },
      ],
      "list": true,
      "type": "list",
    },
  ],
  "list": true,
  "simple": false,
  "type": "root",
}
`;
